### DISCUSSION TITLE
Query a directory or container's digest

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12521] [Ideas]
It would be useful to be able to query the digest of a directory or container. A typical use case would be to use that digest as a tag when pushing an image to a docker registry - thus ensuring that a given version of an image is only pushed to one tag.

cc @kpenfound

**Comment by mircubed** [comment]
@kpenfound do you happen to know if any users have requested this yet? cc @d3rp3tt3

**Comment by kpenfound** [comment]
@mircubed it's definitely come up, though I can't think of who specifically was asking for it. There are also a number of cases we've seen with users where it would benefit them but they didn't know to ask for this specific solution

**Comment by dbason** [comment]
We would definitely use this.  One example would be producing a canonical manifest of the digests for out application (it's made of multiple images).

**Comment by shykes** [comment]
This is not available yet, but someone mentioned a possible workaround: if the source directory is ingested in a git repo, you can use the git tree id for the target directory. The tree ID will remain stable for a given directory contents, unlike the commit ID.

**Comment by shykes** [comment]
I think is done, right @jedevc ?

**Comment by jedevc** [comment]
`Directory.digest` was added in https://github.com/dagger/dagger/pull/8282, `Container.digest` still doesn't exist, so we should leave this one open.

**Comment by jedevc** [comment]
What should the behavior of `Container.digest` be? Should it just be a content digest of the rootfs? I'm guessing it should probably also include all the labels/annotations/workdir/etc - this is a bit more fiddly.

One note - `Container.digest` probably can't return the digest that would actually be pushed to the container registry - otherwise, we'd have to export the entire container. If that actually *is* the intention, then `Publish` already returns the digest that got published.

**Comment by frantjc** [comment]
@jedevc IMO it should be the the digest that would actually be pushed to the container registry. It would be nice to be able to get the digest without having to push the image somewhere, and the digest of the rootfs is already pretty easily available via `container | rootfs | digest`, anyway.

I see an [issue in upstream buildkitd](https://github.com/moby/buildkit/issues/4854) that implies that it is not possible to _just_ export the manifest of an image, but I see [code in Dagger](https://github.com/dagger/dagger/blob/5bd224063062f190fbb6cad2417bb51ece481a94/internal/buildkit/exporter/containerimage/writer.go#L524) that seems to imply that it is possible? Not sure if that's something added in your buildkitd fork, if the issue that I linked above is incorrect, or if I'm misreading the code. If it's possible to do that, computing the digest that would get pushed to the container registry is straight-forward.

We'd just need to figure out how to get from:

```go
// core/container.go
// ...
func (container *Container) Digest (ctx context.Context) (string, error) {
  // ...
}
```

To getting the manifest from `ImageWriter.commitDistributionManifest`.
